driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -0,0 +1,121 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.bson.BsonArray;
+import org.bson.BsonDocument;
+import org.bson.BsonString;
+import org.bson.BsonValue;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.function.Function;
+
+final class MqlExpression<T extends Expression>
+        implements Expression, BooleanExpression, IntegerExpression, NumberExpression,
+        StringExpression, DateExpression, DocumentExpression, ArrayExpression<T> {
+
+    private final Function<CodecRegistry, BsonValue> fn;
+
+    MqlExpression(final Function<CodecRegistry, BsonValue> fn) {
+        this.fn = fn;
+    }
+
+    /**
+     * Exposes the evaluated BsonValue so that expressions may be used in
+     * aggregations. Non-public, as it is intended to be used only by the
+     * {@link MqlExpressionCodec}.
+     */
+    BsonValue toBsonValue(final CodecRegistry codecRegistry) {
+        return fn.apply(codecRegistry);
+    }
+
+    private Function<CodecRegistry, BsonValue> astDoc(final String name, final BsonDocument value) {
+        return (cr) -> new BsonDocument(name, value);
+    }
+
+    private Function<CodecRegistry, BsonValue> ast(final String name) {
+        return (cr) -> new BsonDocument(name, this.toBsonValue(cr));
+    }
+
+    private Function<CodecRegistry, BsonValue> ast(final String name, final Expression param1) {
+        return (cr) -> {
+            BsonArray value = new BsonArray();
+            value.add(this.toBsonValue(cr));
+            value.add(extractBsonValue(cr, param1));
+            return new BsonDocument(name, value);
+        };
+    }
+
+    private Function<CodecRegistry, BsonValue> ast(final String name, final Expression param1, final Expression param2) {
+        return (cr) -> {
+            BsonArray value = new BsonArray();
+            value.add(this.toBsonValue(cr));
+            value.add(extractBsonValue(cr, param1));
+            value.add(extractBsonValue(cr, param2));
+            return new BsonDocument(name, value);
+        };
+    }
+
+    /**
+     * Takes an expression and converts it to a BsonValue. MqlExpression will be
+     * the only implementation of Expression and all subclasses, so this will
+     * not mis-cast an expression as anything else.
+     */
+    private static BsonValue extractBsonValue(final CodecRegistry cr, final Expression expression) {
+        return ((MqlExpression<?>) expression).toBsonValue(cr);
```

Comment:
I'm surprised this doesn't issue an unchecked warning, I don't get why it doesn't?!

Reply:
The [`unchecked`](https://docs.oracle.com/en/java/javase/19/docs/specs/man/javac.html#:~:text=Gives%20more%20detail%20for%20unchecked%20conversion%20warnings%20that%20are%20mandated%20by%20the%20Java%20Language%20Specification) `-Xlint` `javac` key deals only with the unchecked narrowing conversions. The JLS states that [_"A narrowing reference conversion from a type `S` to a parameterized class or interface type `T` is unchecked, unless at least one of the following is true: All of the type arguments of `T` are unbounded wildcards. ..."_](https://docs.oracle.com/javase/specs/jls/se19/html/jls-5.html#jls-5.1.6.2).

It's easy to see why the above is fine: a paremetrized type with all type arguments being unbounded wildcards is a type for which the program has no expectation on the type arguments, so there is nothing that `javac` was required to check but was unable to.

Reply:
Thanks :)

Reply:
While I was away (if you know what I [mean](https://www.google.com/search?q=toilet&oq=toilet&aqs=chrome..69i57.810j0j7&sourceid=chrome&ie=UTF-8)), I realized that the argument _"the program has no expectation on the type arguments"_ is applicable to raw types just as it is applicable to types parametrized with `?`. However, `javac` warns us about raw `MqlExpression` and does not warn when we cast to `MqlExpression<?>`.

While I can't give a formal explanation, I have a hand-weaving explanation via an example. However, I will have to use `List<String>` instead of `Expression` and `ArrayList<?>` instead of `MqlExpression<?>` because there is no practical problem with `MqlExpression` in the code we are discussing, but `javac` cannot know that, and warns us by following a simple algorithm.

The problem with rawtypes (`ArrayList`), which we don't have with `?`-parametrized types (`ArrayList<?>`) and therefore, there is no need for a compiler to warn us, is demonstrated with the following example:

```java
List<String> strings = new ArrayList<>();
ArrayList objects = (ArrayList) strings; // javac warning: [rawtypes] found raw type: ArrayList
// pollute the heap
objects.add(42); // javac warning: [unchecked] unchecked call to add(E) as a member of the raw type ArrayList
// enjoy the consequences
strings.get(0); // java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.String
```

the same example won't compile if we use `ArrayList<?>`:

```java
List<String> strings = new ArrayList<>();
ArrayList<?> objects = (ArrayList<?>) strings;
// try to pollute the heap
objects.add(42); // javac error: incompatible types: int cannot be converted to CAP#1 ... where ... CAP#1 extends Object from capture of ?
// there are no consequences as the code does not compile
strings.get(0);
```

So not only a `?`-paremetrized type express no expectation on the type arguments, but also makes the values of the type kinda sorta "unmodifiable" (I have no idea how to formally express this property), which is how it is different from raw types.